feature(cluster.py): support a new fuzzing tool: cassandra-harry

amoskong opened the pull request:
Ticket: https://trello.com/c/tFSmBovn/2395-explore-cassandra-harry-new-fuzzing-tool-from-cassandra

    This patchset added a new kind of workload.
    
    /usr/bin/cassandra-harry on loader points a wrapper script CLI, it accepts
    `-node` and `-run-time` arguments.


## PR pre-checks (self review)
<!--- PR should be created as Draft, when CI finished and relevant checkboxes selected, add reviewers and then click on "Ready for review" button.-->
<!--- Put an `x` in all the boxes that apply or create PR and then click on all relevant checkboxes: -->
- [x] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [best practices](https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI)
- [x] I didn't leave commented-out/debugging code
- [x] I added the relevant `backport` labels
- [ ] New configuration option are added and documented (in `sdcm/sct_config.py`)
- [ ] I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)
- [ ] All new and existing unit tests passed (CI)
- [ ] I have updated the Readme/doc folder accordingly (if needed)

fruch commented on sdcm/cassandra_harry_thread.py:
```diff
@@ -0,0 +1,184 @@
+                        event.add_info(node=self.node, line=line, line_number=line_number).publish()
+
+
+class CassandraHarryThread:
```
I would recommend looking at YcsbThread code, and base it on docker.
It would be easier to test it, and no need of pre-install anything into the AMIs

amoskong replied:
That's even better, I will try. Thanks.

roydahan replied:
For now I'm ok not investing on making it like that and not as docker.
we install it on the fly.
later if we see some fruits out of it, we can move it to docker (or while we move everything).

roydahan commented:
Add a jenkins job to run it in staging.
After we merge it, copy the job to master and change it to run from master branch.

roydahan commented:
Please add a screenshot of the monitor to see if it's doing anything :)

roydahan commented on test-cases/longevity/longevity-harry-2h.yaml:
```diff
@@ -0,0 +1,16 @@
+n_loaders: 2
+n_monitor_nodes: 1
+
+instance_type_db: 'i3.4xlarge'
```
Try to use the minimal machine type, even i3.large.
(unless it's really stressing).

roydahan commented:
@amoskong what's the status of this?
Where is your trello task for it?

Let's fix it and merge it.

bentsi commented on sdcm/cassandra_harry_thread.py:
```diff
@@ -0,0 +1,184 @@
+                if self.stopped():
+                    break
+
+                for pattern, event in SCYLLA_BENCH_ERROR_EVENTS_PATTERNS:
```
why do you search for Scylla Bench error event patterns?
whether remove it or add error patterns that are relevant for the C-H

amoskong replied:
Ok

amoskong commented:
Updated PR: https://github.com/scylladb/scylla-cluster-tests/pull/3159

- add parse of complete

----

- BYO test: https://jenkins.scylladb.com/job/scylla-staging/job/amos/job/longevity-harry-2h-test-amos/9/console
- 10 mins

< t:2021-06-01 22:30:47,348 f:file_logger.py  l:80   c:sdcm.sct_events.file_logger p:INFO  > 2021-06-01 22:30:47.346: (CassandraHarryEvent Severity.NORMAL): type=start node=Node longevity-10gb-3h-harry-pr-loader-node-37fc2927-1 [13.53.91.125 | 10.0.3.225] (seed: False) stress_cmd=/usr/bin/cassandra-harry -run-time 10 -run-time-unit MINUTES

< t:2021-06-01 22:40:53,236 f:file_logger.py  l:80   c:sdcm.sct_events.file_logger p:INFO  > 2021-06-01 22:40:53.234: (CassandraHarryEvent Severity.NORMAL): type=finish node=Node longevity-10gb-3h-harry-pr-loader-node-37fc2927-1 [13.53.91.125 | 10.0.3.225] (seed: False) stress_cmd=/usr/bin/cassandra-harry -run-time 10 -run-time-unit MINUTES

roydahan commented:
@bentsi can you please review as well?

dkropachev commented:
@amoskong , please rebase it

amoskong commented on sdcm/loader.py:
```diff
@@ -193,3 +193,31 @@ def skip_line(self, line) -> bool:
+        pass
+
+    def skip_line(self, line) -> bool:
+        return not 'Reorder buffer size has grown up to' in line
```
Yes, we can only get metrics from this line.
But the other log can still be saved to the log.

amoskong commented:
> @amoskong , please rebase it

Done.

roydahan commented on sdcm/cluster.py:
```diff
@@ -4492,6 +4493,7 @@ def node_setup(self, node, verbose=False, db_node_address=None, **kwargs):  # py
 
         if 'scylla-bench' in self.params.list_of_stress_tools:
             self.install_scylla_bench(node)
+            self.install_cassandra_harry(node)
```
How is it related to scylla-bench?
Why would we want to  install cassandra-harry everytime we have scylla-bench?

We need to  install it only when we need to.

amoskong replied:
fixed.  only install/update cassandra-harry when it exists in the stress_cmd

roydahan commented on sdcm/cassandra_harry_thread.py:
```diff
@@ -0,0 +1,191 @@
+        return self
+
+    @staticmethod
+    def _parse_harry_summary(lines):  # pylint: disable=too-many-branches
```
Please add in comment a sample of the output from the tool.
Is it only completed or failed?

amoskong replied:
added, currently we only check if it's completed or not.
It doesn't output any summary of the write/read ops.

dkropachev commented on longevity_test.py:
```diff
@@ -51,6 +51,8 @@ def _run_all_stress_cmds(self, stress_queue, params):
                                                            stats_aggregate_cmds=False,
                                                            round_robin=self.params.get('round_robin')))
+            elif stress_cmd.startswith('cassandra-harry'):
+                stress_queue.append(self.run_stress_thread(**params))
```
why it is needed ?

roydahan replied:
@amoskong it's exact same line as in the "else"

dkropachev commented on sdcm/cassandra_harry_thread.py:
```diff
@@ -0,0 +1,199 @@
+
+    def _run_stress_harry(self, node, loader_idx, stress_cmd, node_list):
+
+        CassandraHarryEvent.start(node=node, stress_cmd=stress_cmd).publish()
```
Please take a look at stress_thread and do it in the same way

dkropachev commented on sdcm/cluster.py:
```diff
@@ -4704,6 +4709,24 @@ def install_scylla_bench(node):
+            """))
+        node.remoter.run(shell_script_cmd("""\
+            rm -rf cassandra-harry
+            git clone -b scylla-branch https://github.com/amoskong/cassandra-harry
```
hook it to specific version/commit_id

roydahan replied:
I think it's not needed, it's better to use latest and if needed to fix it.

dkropachev replied:
@roydahan , we have had same issue with all of the loaders, one day it is getting updated without us noticing it, test begin to fail, no one knows what is going on, after hours/days of investigation it turns out that it is just a tool got upgraded.

roydahan commented:
This is the official job but still run from a private branch (amoskong):
https://jenkins.scylladb.com/job/scylla-master/job/longevity/job/longevity-harry-2h-test/

roydahan commented on sdcm/cluster.py:
```diff
@@ -4704,6 +4709,24 @@ def install_scylla_bench(node):
+            """))
+        node.remoter.run(shell_script_cmd("""\
+            rm -rf cassandra-harry
+            git clone -b scylla-branch https://github.com/amoskong/cassandra-harry
```
Please also ask Avi to move it into scylladb repos and change it here from "amosking" to scylladb.
Another option, is trying to contribute our changes to original apache repo.
Last option is use the original repo and apply the patches here (can break easily with changes but will be most up to date).

roydahan replied:
But this can be in a following PR

enaydanov commented on test-cases/longevity/longevity-harry-2h.yaml:
```diff
@@ -0,0 +1,21 @@
+nemesis_interval: 2
+ssh_transport: 'libssh2'
+
+user_prefix: 'longevity-10gb-3h'
```
Need to change user_prefix

enaydanov commented on test-cases/longevity/longevity-harry-2h.yaml:
```diff
@@ -0,0 +1,21 @@
+user_prefix: 'longevity-10gb-3h'
+space_node_threshold: 64424
+use_mgmt: false
+#reuse_cluster: 37fc2927-1bfa-48ba-9350-4bd5a51b5000
+#reuse_cluster: b9dcf778-1164-4227-9c04-88db64a1448d
+reuse_cluster: 78b2c81b-43e7-4340-8686-64fcda1a3f44
```
Some debug stuff? Need to remove.

amoskong replied:
Oh, my mistake

enaydanov commented on sdcm/tester.py:
```diff
@@ -1368,6 +1370,9 @@ def run_stress_thread(self, stress_cmd, duration=None, stress_num=1, keyspace_nu
         elif stress_cmd.startswith('scylla-bench'):
             params['stop_test_on_failure'] = stop_test_on_failure
             return self.run_stress_thread_bench(**params)
+        elif stress_cmd.startswith('cassandra-harry') or 'cassandra-harry' in stress_cmd:
```
```python
if 'cassandra-harry' in stress_cmd:
    ...
```
is enough here (it covers `.startswith()` case too)

amoskong replied:
You are right. the startswith is unnecessary

enaydanov commented on sdcm/tester.py:
```diff
@@ -1439,6 +1444,33 @@ def run_stress_thread_bench(self, stress_cmd, duration=None, stress_num=1, keysp
+        )
+        harry_thread.run()
+        scylla_encryption_options = self.params.get('scylla_encryption_options')
+        if scylla_encryption_options and 'write' in stress_cmd:
```
Is it a case for Harry to have `write` substring in `stress_cmd` ?

amoskong replied:
No, so it's a problem.

eg: /usr/bin/cassandra-harry -run-time 2 -run-time-unit HOURS

But we may need to add support of assigning read/write opts in cassandra-harry cmdline, the wrap script was added by myself.

enaydanov commented on sdcm/tester.py:
```diff
@@ -1439,6 +1444,33 @@ def run_stress_thread_bench(self, stress_cmd, duration=None, stress_num=1, keysp
+        if self.create_stats:
+            self.update_stress_cmd_details(stress_cmd, stresser="cassandra-harry", aggregate=stats_aggregate_cmds)
+        harry_thread = CassandraHarryThread(
+            stress_cmd, loader_set=self.loaders, timeout=timeout,
```
Please, put one parameter per line.

amoskong replied:
fixed.

enaydanov commented on sdcm/tester.py:
```diff
@@ -1439,6 +1444,33 @@ def run_stress_thread_bench(self, stress_cmd, duration=None, stress_num=1, keysp
+                                stop_test_on_failure=True):  # pylint: disable=too-many-arguments,unused-argument
+
+        timeout = self.get_duration(duration)
+        stop_test_on_failure = False if not self.params.get("stop_test_on_stress_failure") else stop_test_on_failure
```
Do you find following code is more clear for this line, don't you?
```python
if not self.params.get("stop_test_on_stress_failure"):
    stop_test_on_failure = False
```

amoskong replied:
Yes

enaydanov commented on sdcm/sct_events/base.py:
```diff
@@ -52,13 +52,14 @@ def __init__(self, severities_conf: str = DEFAULT_SEVERITIES):
     def __setitem__(self, key: str, value: Type[SctEvent]):
         if not value.is_abstract() and key not in self.max_severities:
-            raise ValueError(f"There is no max severity configured for {key}")
+            raise ValueError(f"{value} There is no max severity configured for {key} {self.max_severities.keys()}")
```
Is it some debugging leftover here?

amoskong replied:
Yes, removed.

enaydanov commented on sdcm/loader.py:
```diff
@@ -188,3 +190,32 @@ def skip_line(self, line) -> bool:
+    def skip_line(self, line) -> bool:
+        return not 'Reorder buffer size has grown up to' in line
+
+    @staticmethod
+    def split_line(line: str) -> list:
+        return [element.strip() for element in line.split()]
```
Just `.split()` do the same:

```python
>>> '   1    2   '.split()
['1', '2']
```

amoskong replied:
fixed.

enaydanov commented on sdcm/db_stats.py:
```diff
@@ -136,6 +136,21 @@ def get_stress_bench_cmd_params(cmd):
+    cmd = cmd.strip().split('cassandra-harry')[1].strip()
+    cmd_params = {}
+    for key in ['run-time', 'run-time-unit']:
+        match = re.search(r'(-' + key + r'\s+([^-| ]+))', cmd)
```
Please, use f-string instead of concatenation.

enaydanov commented on sdcm/db_stats.py:
```diff
@@ -136,6 +136,21 @@ def get_stress_bench_cmd_params(cmd):
+    :param cmd: stress cmd
+    :return: dict with params
+    """
+    cmd = cmd.strip().split('cassandra-harry')[1].strip()
```
First `.strip()` is redundant here.

amoskong replied:
fixed.

roydahan commented on sdcm/cluster.py:
```diff
@@ -383,6 +384,25 @@ def install_scylla_bench(self):
+            """))
+        self.remoter.run(shell_script_cmd("""\
+            rm -rf cassandra-harry
+            git clone -b scylla-branch https://github.com/amoskong/cassandra-harry
```
Amos sent a PR to apache, but for now we can use his fork till they merge it.
I'll create a task reminder for that.

roydahan commented:
@amoskong please go over the comments from Dmitry and Evgeniy ASAP.

amoskong commented:
> @amoskong please go over the comments from Dmitry and Evgeniy ASAP.

I updated the PR 3 days ago, and addressed all comments. Is there something missing?

dkropachev commented:
@amoskong , this one is not fixed - https://github.com/scylladb/scylla-cluster-tests/pull/3159/files#r671510133

roydahan commented:
@dkropachev I think we need to merge it as long as we don't have conflicts and then send some fixes for it.
(@amoskong officially left, so he can contribute only on his spare time).

dkropachev commented:
> @dkropachev I think we need to merge it as long as we don't have conflicts and then send some fixes for it. (@amoskong officially left, so he can contribute only on his spare time).

Sure lets do that and i will create trello tasks to fix things.

amoskong commented:
Thanks @dkropachev
